text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+situations, it is by no means a complete solution.
+
+[`readonly` crate]: https://crates.io/crates/readonly
+
+### Where does a mutation occur?
+
+There is one major question: what even counts as a mutation? This is not as straightforward as you
+might think. If you write
+
+```rust
+let mut x = 5;
+let y = &mut x;
+*y = 6;
+```
+
+It is without question that a mutation occurs. But where? Does it occur on the second or third line?
+In this example, it would not matter, but it is easy to imagine passing a mutable reference to a
+function that then mutates the value. There, it is not clear where the mutation occurs. The answer
+is that the mutation occurs on the line where the reference is taken. This is the choice that makes
+the most sense from the perspective of the user.
+
+```rust
+fn foo<T>(x: &mut T, value: T) {
+    if random() {
+        *x = value;
+    }
+}
+
+let mut x = 5;
+foo(&mut x, 6);
+```
+
+Here, `x` is mutably borrowed on the final line, but the value is changed in memory inside the `if`
+block. You might say, logically, that the mutation occurs inside the `if` block. But if we use this
+definition, then we could not know about the mutation until after monomorphization. Errors generated
+post-monomorphization are generally frowned upon, as it happens quite late in the compilation
+process. But consider this: what if `x` is not actually mutated within the body of `foo`? Now we
+have a window into what actually happens inside the function, and it is something that is not stated
+in the function signature. Not great. In this specific example, it is not even deterministic!
+Because of this, it is quite literally impossible to know whether `x` is actually mutated inside a
+given function. As a result we have no choice: the error _must_ be generated at the point where the
+reference is taken.
+
+Okay, we solved that problem. We know that the mutable use happens on the final line. But what about
+this?
+
+```rust
+let x = Cell::new(5);
+x.set(6);
+```
+
+Rust, as you may know, has [interior mutability]. This is what we are using here. `x` is not
+declared mutable, and it does not need to be. This is the beauty of interior mutability. But it
+introduces a key question: where is the mutation? The answer is that it is **not** a mutation for
+the purposes of this restriction. This is not because the value is not changed: it is. Rather, it is
+the logical result of the semantics of `mut` restrictions and where errors must occur (as described
+after the previous example). If errors are emitted at the point where the mutable reference is
+created, then there can be no such error here, as no mutable reference is ever created. `Cell::set`
+is a method that takes `&self`, not `&mut self`. Surely we can make interior mutability a special
+case, right? No; the only way to work around this is to make any reference to a type with interior
+mutability considered a mutation. Consequently, you could never have a reference to a type
```

Comment:
```suggestion
is a method that takes `&self`, not `&mut self`. Interior mutability is not a special-case; the only way to work around this would be to make even non-mutable references to a type with interior
mutability considered a mutation. Consequently, you could never have a reference to a type
```